EngineLayer/DecoyOnTheFly.cs

```diff
@@ -0,0 +1,168 @@
+                    }
+                    newBase[fillPosition] = evaporatingBase[extractPosition];
+                    revisedAminoAcidOrder[fillPosition] = extractPosition;
+                    if (pwsm.AllModsOneIsNterminus.ContainsKey(extractPosition + 2))
+                    {
+                        newModificationsDictionary.Add(fillPosition + 2, pwsm.AllModsOneIsNterminus[extractPosition + 2]);
+                    }
+                    fillPosition++;
+                }
+                extractPosition--;
+            }
+
+            string newBaseString = new string(newBase);
+
+            var proteinSequence = pwsm.Protein.BaseSequence;
+            var aStringBuilder = new StringBuilder(proteinSequence);
+            aStringBuilder.Remove(pwsm.OneBasedStartResidueInProtein - 1, pwsm.BaseSequence.Length);
+            aStringBuilder.Insert(pwsm.OneBasedStartResidueInProtein - 1, newBaseString);
+            proteinSequence = aStringBuilder.ToString();
+
+            Protein decoyProtein = new Protein(proteinSequence, "DECOY_" + pwsm.Protein.Accession, null, new List<Tuple<string, string>>(), new Dictionary<int, List<Modification>>(), null, null, null, true);
+
+            DigestionParams d = pwsm.DigestionParams;
+
+            if (newBaseString != pwsm.BaseSequence)
+            {
+                return new PeptideWithSetModifications(decoyProtein, d, pwsm.OneBasedStartResidueInProtein, pwsm.OneBasedEndResidueInProtein, pwsm.CleavageSpecificityForFdrCategory, pwsm.PeptideDescription, pwsm.MissedCleavages, newModificationsDictionary, pwsm.NumFixedMods, newBaseString);
+            }
+            else
+            {
+                //The reverse decoy procedure failed to create a PeptideWithSetModificatons with a different sequence. Therefore,
+                //we retrun the mirror image peptide.
+                return GetPeptideMirror(pwsm, revisedAminoAcidOrder);
+            }
+
+        }
+
+        //Returns a PeptideWithSetModifications mirror image. Used when reverse decoy sequence is same as target sequence
+        public static PeptideWithSetModifications GetPeptideMirror(PeptideWithSetModifications pwsm, int[] revisedOrderNisOne)
+        {
+            Dictionary<int, Modification> newModificationsDictionary = new Dictionary<int, Modification>();
+            //Copy N-terminal modifications from target dictionary to decoy dictionary.
+            if (pwsm.AllModsOneIsNterminus.ContainsKey(1))
+            {
+                newModificationsDictionary.Add(1, pwsm.AllModsOneIsNterminus[1]);
+            }
+
+            //First step is to reverse the position of all modifications except the mod on the peptide N-terminus.
+            if (pwsm.AllModsOneIsNterminus.Any())
+            {
+                int newPosition = pwsm.BaseSequence.Length + 1;
+                for (int i = 2; i < pwsm.BaseSequence.Length + 2; i++)
+                {
+                    if (pwsm.AllModsOneIsNterminus.ContainsKey(i))
```

Comment:
This is going to be slow, since you're looking through each amino acid position for a key. I'd recommend refactoring using a foreach(var kvp in pwsm.AllModsOneIsNterminus). The newPosition should be pwsm.baseSequence.Length -kvp.Key and the pwsm.AllModsOneIsNterminus[i] should be kvp.Value.